src/wayland/mod.rs

```diff
@@ -72,60 +73,119 @@ impl WaylandImpl {
             display,
             surface,
             buffers: Default::default(),
+            size: None,
         })
     }
 
-    fn buffer(&mut self, width: i32, height: i32) -> &WaylandBuffer {
-        self.buffers = Some(if let Some((front, mut back)) = self.buffers.take() {
-            // Swap buffers; block if back buffer not released yet
+    pub fn resize(&mut self, width: u32, height: u32) -> Result<(), SoftBufferError> {
+        self.size = Some(
+            (|| {
+                let width = NonZeroI32::new(i32::try_from(width).ok()?)?;
+                let height = NonZeroI32::new(i32::try_from(height).ok()?)?;
+                Some((width, height))
+            })()
+            .ok_or(SoftBufferError::SizeOutOfRange { width, height })?,
+        );
+        Ok(())
+    }
+
+    pub fn buffer_mut(&mut self) -> Result<BufferImpl, SoftBufferError> {
+        let (width, height) = self
+            .size
+            .expect("Must set size of surface before calling `buffer_mut()`");
+
+        if let Some((_front, back)) = &mut self.buffers {
+            // Block if back buffer not released yet
             if !back.released() {
-                let mut event_queue = self.display.event_queue.lock().unwrap();
+                let mut event_queue = self.display.event_queue.borrow_mut();
```

Comment:
This is behind `&mut`, so you can `get_mut()` directly.

Reply:
You can still do that.

Reply:
It's not especially elegant, but it seems better than having softbuffer create an event queue per surface, and should work well enough in practice.